File: backends/php-laravel/app/Support/Http/Responses/Traits/Pager.php
<?php

/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */

declare(strict_types=1);

namespace App\Support\Http\Responses\Traits;

use App\Support\Enums\SortType;
use App\Support\Traits\Assign;

/**
 * ページャーTraitです。
 *
 * @author Satoshi Nagashiba <nagashibas@sii-japan.co.jp>
 * @package App\Support\Http\Responses\Traits
 */
trait Pager
{
    protected int $count = 0;
    protected int $offset = 0;
    protected int $limit = 0;
    protected string $sort = '';
    protected SortType $sortType = SortType::NONE;
    protected string $sortTypeValue = '';

    /**
     * ページャーを作成します。
     *
     * @param int $recordCount データ件数
     * @return mixed ページャー
     */
    public function createPager(int $recordCount): mixed
    {
        $defaultPageCount = $this->getDefaultPageCount();
        $page = $this->getPage();
        $pageCount = $this->getPageCount();
        $value = [
            'count' => $this->count,
            'limit' => $this->limit,
            'next' => $this->hasNextPage(),
            'previous' => $this->hasPreviousPage(),
            'page' => $page,
            'nextPage' => $page + 1,
            'previousPage' => $page - 1,
            'pageCount' => $pageCount,
            'first' => ($page > 1),
            'last' => ($pageCount > $page),
            'firstRecordCount' => $this->getOffset() + 1,
            'lastRecordCount' => $this->getOffset() + $recordCount,
            'startPage' => $this->getStartPage($defaultPageCount),
            'endPage' => $this->getEndPage($defaultPageCount),
        ];
        $this->sortTypeValue = $this->sortType->value;

        return new class ($value) {
            use Assign;

            public int $count = 0;
            public int $limit = 0;
            public bool $next = false;
            public bool $previous = false;
            public int $page = 0;
            public int $nextPage = 0;
            public int $previousPage = 0;
            public int $pageCount = 0;
            public bool $first = false;
            public bool $last = false;
            public int $firstRecordCount = 0;
            public int $lastRecordCount = 0;
            public int $startPage = 0;
            public int $endPage = 0;

            public function __construct(array $value)
            {
                $this->assign($value);
            }
        };
    }

    /**
     * オフセットを取得します。
     *
     * @return int オフセット
     */
    protected function getOffset(): int
    {
        if (!empty($this->count)) {
            if ($this->offset > $this->getLastPageOffset()) {
                return $this->getLastPageOffset();
            }
        }
        return $this->offset;
    }

    /**
     * 次頁があるか確認します。
     *
     * @return bool 次頁がある場合 true を返します
     */
    protected function hasNextPage(): bool
    {
        return $this->getPageCount() > $this->getPage();
    }

    /**
     * 前頁があるか確認します。
     *
     * @return bool 前頁がある場合 true を返します
     */
    protected function hasPreviousPage(): bool
    {
        return $this->getPage() > 1;
    }

    /**
     * 頁を取得します。
     *
     * @return int 頁
     */
    protected function getPage(): int
    {
        if (empty($this->limit)) {
            return 1;
        }
        return floor(ceil($this->getOffset() / $this->limit) + 1);
    }

    /**
     * 開始頁を取得します。
     *
     * @param int $defaultPageCount デフォルト頁数
     * @return int 開始頁
     */
    protected function getStartPage(int $defaultPageCount): int
    {
        $start = $this->getPage() - ($defaultPageCount - 1);
        if ($start <= 0) {
            $start = 1;
        }
        return $start;
    }

    /**
     * 最終頁を取得します。
     *
     * @param int $defaultPageCount デフォルト頁数
     * @return int 最終頁
     */
    protected function getEndPage(int $defaultPageCount): int
    {
        $end = $this->getStartPage($defaultPageCount) + ($defaultPageCount - 1);
        $pageCount = $this->getPageCount();
        if ($end > $pageCount) {
            $end = $pageCount;
        }
        return $end;
    }

    /**
     * 頁数を取得します。
     *
     * @return int 頁数
     */
    protected function getPageCount(): int
    {
        if (empty($this->limit)) {
            return 1;
        }

        $result = ceil($this->count / $this->limit);
        if ($result < 1) {
            $result = 1;
        }

        return $result;
    }

    /**
     * 最終頁オフセットを取得します。
     *
     * @return int 最終頁オフセット
     */
    protected function getLastPageOffset(): int
    {
        return ($this->getPageCount() * $this->limit) - $this->limit;
    }

    /**
     * デフォルト頁数を取得します。
     *
     * @return int デフォルト頁数
     */
    protected function getDefaultPageCount(): int
    {
        return config('authorization.app.page.count');
    }
}

File: backends/go-gin/internal/handler/pager.go
package handler

import "math"

// defaultPageCount はページャー番号ボタンの表示数です。
const defaultPageCount = 5

// Pager はサーバーサイドページングのレスポンス構造体です。
type Pager struct {
	Count            int  `json:"count"`
	Limit            int  `json:"limit"`
	Next             bool `json:"next"`
	Previous         bool `json:"previous"`
	Page             int  `json:"page"`
	NextPage         int  `json:"nextPage"`
	PreviousPage     int  `json:"previousPage"`
	PageCount        int  `json:"pageCount"`
	First            bool `json:"first"`
	Last             bool `json:"last"`
	FirstRecordCount int  `json:"firstRecordCount"`
	LastRecordCount  int  `json:"lastRecordCount"`
	StartPage        int  `json:"startPage"`
	EndPage          int  `json:"endPage"`
}

// BuildPager はページング情報を構築します。
func BuildPager(count, limit, offset, recordCount int) Pager {
	if limit <= 0 {
		limit = 20
	}

	pageCount := int(math.Ceil(float64(count) / float64(limit)))
	if pageCount < 1 {
		pageCount = 1
	}

	lastPageOffset := (pageCount * limit) - limit
	if count > 0 && offset > lastPageOffset {
		offset = lastPageOffset
	}

	page := int(math.Floor(math.Ceil(float64(offset)/float64(limit)))) + 1

	startPage := page - (defaultPageCount - 1)
	if startPage <= 0 {
		startPage = 1
	}
	endPage := startPage + (defaultPageCount - 1)
	if endPage > pageCount {
		endPage = pageCount
	}

	return Pager{
		Count:            count,
		Limit:            limit,
		Next:             pageCount > page,
		Previous:         page > 1,
		Page:             page,
		NextPage:         page + 1,
		PreviousPage:     page - 1,
		PageCount:        pageCount,
		First:            page > 1,
		Last:             pageCount > page,
		FirstRecordCount: offset + 1,
		LastRecordCount:  offset + recordCount,
		StartPage:        startPage,
		EndPage:          endPage,
	}
}
